File: iosApp/iosApp/Core/Util/CastScheduleAttendanceUtils.swift
//
//  CastScheduleAttendanceUtils.swift
//  ConCafe
//
//  Created by Codex on 2026/04/10.
//

import Foundation
import Shared

final class CastScheduleAttendanceUtils {
    static func todaySchedule(
        from schedules: [CastSchedule],
        currentDate: String = TimeUtils.currentIsoDate()
    ) -> CastSchedule? {
        schedules.first { $0.date == currentDate }
    }

    static func todayAttendanceStatus(
        schedules: [CastSchedule],
        currentDate: String = TimeUtils.currentIsoDate(),
        currentMinutes: Int = TimeUtils.currentTimeMinutes()
    ) -> CastAttendanceStatus {
        attendanceStatus(
            schedule: todaySchedule(from: schedules, currentDate: currentDate),
            currentMinutes: currentMinutes
        )
    }

    static func attendanceStatus(
        schedule: CastSchedule?,
        currentMinutes: Int = TimeUtils.currentTimeMinutes()
    ) -> CastAttendanceStatus {
        guard let schedule else {
            return .off
        }

        let startMinutes = TimeUtils.parseTimeMinutes(schedule.startTime)
        let endMinutes = TimeUtils.parseTimeMinutes(schedule.endTime)
        if currentMinutes < startMinutes {
            return .upcoming
        }
        if currentMinutes <= endMinutes {
            return .onShift
        }
        return .completed
    }
}

File: composeApp/src/commonMain/kotlin/com/hhp227/concafe/core/util/CastScheduleAttendanceUtils.kt
package com.hhp227.concafe.core.util

import com.hhp227.concafe.domain.model.CastAttendanceStatus
import com.hhp227.concafe.domain.model.CastSchedule

object CastScheduleAttendanceUtils {
    fun todaySchedule(
        schedules: List<CastSchedule>,
        currentDate: String = TimeUtils.currentIsoDate()
    ): CastSchedule? {
        return schedules.firstOrNull { schedule -> schedule.date == currentDate }
    }

    fun todayAttendanceStatus(
        schedules: List<CastSchedule>,
        currentDate: String = TimeUtils.currentIsoDate(),
        currentMinutes: Int = TimeUtils.currentTimeMinutes()
    ): CastAttendanceStatus {
        return attendanceStatus(
            schedule = todaySchedule(schedules, currentDate),
            currentMinutes = currentMinutes
        )
    }

    fun attendanceStatus(
        schedule: CastSchedule?,
        currentMinutes: Int = TimeUtils.currentTimeMinutes()
    ): CastAttendanceStatus {
        if (schedule == null) {
            return CastAttendanceStatus.OFF
        }

        val startMinutes = TimeUtils.parseTimeToMinutes(schedule.startTime)
        val endMinutes = TimeUtils.parseTimeToMinutes(schedule.endTime)
        return when {
            currentMinutes < startMinutes -> CastAttendanceStatus.UPCOMING
            currentMinutes <= endMinutes -> CastAttendanceStatus.ON_SHIFT
            else -> CastAttendanceStatus.COMPLETED
        }
    }
}
